geth/jail/cell_test.go

```diff
@@ -91,43 +98,83 @@ func (s *CellTestSuite) TestCellFetchRace() {
 
 	err := cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
 	s.NoError(err)
+
 	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
 	s.NoError(err)
 
-	// run JS code for fetching valid URL
-	_, err = cell.Run(`fetch('` + server.URL + `').then(function(r) {
-		return r.text()
+	fetchCode := `fetch('%s?i=%d').then(function(r) {
+ 		return r.text()
 	}).then(function(data) {
 		__captureSuccess(data)
 	}).catch(function (e) {
 		__captureError(e)
-	})`)
+	})`
+
+	for i := 0; i < requestCount; i++ {
+		_, err = cell.Run(fmt.Sprintf(fetchCode, server.URL, i))
+		s.NoError(err)
+	}
+
+	for i := 0; i < requestCount; i++ {
+		select {
+		case data := <-dataCh:
+			// Mark the request as successful.
+			expected[data.String()] = true
+		case _ = <-errCh:
+			s.Fail("fetch failed to complete the request")
+		case <-time.After(5 * time.Second):
+			s.Fail("test timed out")
+			return
+		}
+	}
+
+	// Make sure every request was completed successfully.
+	for i := 0; i < requestCount; i++ {
+		s.Equal(expected[fmt.Sprintf("%d", i)], true)
+	}
+
+	// There might be some tasks about to call `ready`,
+	// add a little delay before `TearDownTest` closes the loop.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Is there any way to make it actually wait for all the tasks to be `ready`?
I afraid that this timeout might well differ on your machine and, say in Travis, where everything is super slow...

Reply:
@mandrigin the problem causes from the lock in this code; https://github.com/status-im/status-go/blob/develop/geth/jail/internal/loop/loop.go#L133-L135

In some cases there might be a task that passed the lock but not sent to `l.ready` channel yet when `l.ready` channel was already closed by `TearDownTest`. I think 100ms is plenty of time to make sure the channel is alive until the leftover task is passed through...

Reply:
ah, ok
@divan what do you think?

Reply:
i experimented a bit with this code: https://github.com/dshulyak/status-go/commit/29e0d5100ddffd387bf10d5d9117536cd4f43a39

with those changes there are no races. but i will probably open another review.
it is probably ok to leave this timeout temporary

Reply:
Nice! 👍

Reply:
opened PR https://github.com/status-im/status-go/pull/632

Reply:
Yeah, I think this timeout is fine for now here. Also, keep in mind the real-world usage patterns of Jail and fetch. We have only a couple of dapps that use fetch afaik.